Lib/pathlib/_abc.py

```diff
@@ -786,11 +793,11 @@ def glob(self, pattern, *, case_sensitive=None, follow_symlinks=None):
                 # regex filtering, provided we're treating symlinks consistently.
                 if follow_symlinks is not None:
                     while stack and stack[-1] not in specials:
-                        filter_paths = True
-                        stack.pop()
+                        part += sep + stack.pop()
 
                 dir_only = bool(stack)
-                paths = _select_recursive(paths, dir_only, follow_symlinks)
+                match = _compile_pattern(part, sep, case_sensitive) if part != '**' else None
```

Comment:
```suggestion
                # Pass in a pattern to eagerly exclude paths we know don't match.
                # Otherwise, they'll end up in the final set and be filtered anyway.
                match = _compile_pattern(part, sep, case_sensitive) if part != '**' else None
```